### PR TITLE
[Documentation]: Move project structure into its own file.

### LINKED ISSUE #17: Write documentation on Annotation processor

### PR DESCRIPTION
**Pull request by miguelandres**
README.md has grown out of control, it has both too much and too little information. High-level documentation is lacking because it cannot sensibly fit into one giant Markdown file, but at the same time we try to address many different things in the readme.

This is the first out of several CLs in which I plan to split the following parts of the README into their own files:
- Project Structure (**this CL**, it also adds a high level summary for each project and their key pieces)
- Getting Started and dependencies.
- Remixer concepts and supported variable types
- UI configuration (multiple ways to load the ui, these docs do not exist yet)
- Sync and storage configuration (these docs do not exist yet)
- Configuring the Firebase remote controller (still not done)
- Extending remixer (it's possible but not even discussed in current docs)

Fixes #17.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -252,7 +233,7 @@ Yes, the [material-foundation](https://github.com/material-foundation) organizat
 
 ## Contributing
 
-We gladly welcome contributions! If you have found a bug, have questions, or wish to contribute, please follow our [Contributing Guidelines](https://github.com/material-foundation/material-remixer-android/blob/develop/CONTRIBUTING.md).
+We gladly welcome contributions! If you have found a bug, have questions, or wish to contribute, please follow our [Contributing Guidelines](CONTRIBUTING.md) and read the [High-level Project Structure](PROJECT_STRUCTURE.md).
```
Comment (pingpongboss) [inline README.md:236]: your README could server as a better-formatted table of contents. If you plan to have many of these files, you may want to just list them in bullet point format.

File: PROJECT_STRUCTURE.md [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+- The `com.google.android.libraries.remixer.Remixer` class which aggregates all variables and `package com.google.android.libraries.remixer.sync` which takes care of keeping values in sync across different contexts.
+  - `com.google.android.libraries.remixer.sync.LocalValueSyncing` is a non-persistent version of a `com.google.android.libraries.remixer.sync.SynchronizationMechanism`. Persistent versions are available in the `remixer_storage` submodule, those depend on android-specific classes which are unavailable on this project, because of the reasons explained in the next section.
+- Declaration of all annotations exposed by Remixer `package com.google.android.libraries.remixer.annotation` and the annotation binding classes.
+
+The annotation-processing library, `remixer_annotation`, contains code that should only be run by the Java Compiler (and should never be in the classpath for an APK), this code generates Remixer variables from annotations in `com.google.android.libraries.remixer.annotation`.
+
+
+### Constraints on dependencies
+
+**TL;DR: If you contribute to Remixer, you need to know you cannot depend on any android classes in the `remixer_core` submodule**
```
Comment (pingpongboss) [inline PROJECT_STRUCTURE.md]: this sentence is a little ambiguous. Do you mean:
"If you contribute to Remixer, you need to know that the `remixer_core` submodule cannot depend on any android classes"

File: PROJECT_STRUCTURE.md [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+Remixer uses annotations heavily, and annotation processing on Android is very difficult. The Android SDK does not include classes required for annotation processing, therefore if you depend on the Android SDK, you cannot use annotation processing. For this reason there are two submodules that are Java libraries instead of Android libraries, `remixer_core` and `remixer_annotation`.
+
+This means that the annotation processing code (in `remixer_annotation`) is unable to generate code that uses any Android class (i.e. can't generate `OnClickListener`s, or use `Actvity` in generated code, etc.) [^generatingtestedcode]. Furthermore, in some cases, where Remixer needs a Context (which should naturally be an instance of `android.app.Activity` or at least `android.content.Context`) we take `java.lang.Object` instead. It is generally very well explained in comments that this is the case.
```
Comment (pingpongboss) [inline PROJECT_STRUCTURE.md]: Is `[^generatingtestedcode]` supposed to be a footnote? It's not formatted at all in the Preview. You may want to stick with `[1]` which does work.

File: PROJECT_STRUCTURE.md [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+## Storage submodule.
+
+The `remixer_storage` submodule currently contains two implementations of `remixer_core`'s `com.google.android.libraries.remixer.sync.SynchronizationMechanism`:
+- `com.google.android.libraries.remixer.storage.LocalStorage` writes to a shared preferences file each time a value is modified. It uses `SharedPreferences.Editor#apply()` to commit the changes, which means that if there are many changes one after the other it would (probably) hit the disk only once, after all of them complete, and it doesn't block the UI thread at all.
+- `com.google.android.libraries.remixer.storage.FirebaseRemoteController`: <sup id="a2">[2](#f2)</sup> Uses firebase for storage and synchronization back to the device. This is necessary for the upcoming firebase Remote Controller functionality.
+
+## UI submodule
+
+The `remixer_ui` submodule contains the actual Remixer visible UI and a little of the "glue code" that makes everything seamless.
+- `package com.google.android.libraries.remixer.ui.view` contains the classes necessary to use the `RemixerFragment that shows all the widgets that change Remixer.
```
Comment (pingpongboss) [inline PROJECT_STRUCTURE.md]: extra "`"

**Review by pingpongboss — APPROVED** [review]

File: README.md (hunk 1 above)
Comment (miguelandres) [reply README.md:236]: I was thinking of that, but in a bit when I write more of these.

File: PROJECT_STRUCTURE.md (hunk 2 above)
Comment (miguelandres) [reply PROJECT_STRUCTURE.md]: Indeed. THanks.

File: PROJECT_STRUCTURE.md (hunk 3 above)
Comment (miguelandres) [reply PROJECT_STRUCTURE.md]: [1] doesn't work either, I had missed this one when I realised this wasn't working.

File: PROJECT_STRUCTURE.md (hunk 4 above)
Comment (miguelandres) [reply PROJECT_STRUCTURE.md]: Or missing.

**Pull request merged**